client/src/main/java/io/atomix/copycat/client/util/ClientConnection.java

```diff
@@ -185,22 +197,20 @@ public ClientConnection reset(Address leader, Collection<Address> servers) {
    * Connects to the cluster.
    */
   private CompletableFuture<Connection> connect() {
-    CompletableFuture<Connection> future = new CompletableFuture<>();
-
     // If the address selector has been reset then reset the connection.
     if (selector.state() == AddressSelector.State.RESET && connection != null) {
       if (connectFuture != null) {
-        connectFutures.add(future);
         return connectFuture;
       }
 
-      CompletableFuture<Connection> newConnectFuture = new CompletableFuture<>();
-      connectFuture = newConnectFuture;
-      connectFutures.add(future);
-
-      connection.close().whenComplete((result, error) -> connect(newConnectFuture));
+      connectFuture = new OrderedCompletableFuture<>();
+      connectFuture.whenComplete((r, e) -> connectFuture = null);
 
-      return awaitConnect(newConnectFuture);
+      Connection oldConnection = this.connection;
+      this.connection = null;
+      oldConnection.close();
+      connect(connectFuture);
+      return connectFuture;
```

Comment:
The unit tests are failing because of that stupid `CompletableFuture` async call bug. The future returned by `connect()` in `LocalClient` should be completed asynchronously, but it's not because of that bug. It's being completed synchronously, causing `whenComplete` to be called synchronously before the future is ever returned. I hate that bug.